### PR TITLE
Implement optional Gen1 type immunity logic.

### PR DESCRIPTION
**Pull request by GraionDilach**
In Gen1, Pokémons were immune to burn/freeze/paralyze effects of damaging moves if they shared the type with it (this logic is famous for making Tauros immune to Body Slam paralyzation in competetive among others).

I presume this needs tests but I'm not kinda sure what kind of tests should I attempt to write even. I'll proly read some of the other tests to get some ideas.

Relevant Pret reference: https://github.com/pret/pokered/blob/d4d7b91aecf651b06d1f466ecc22b65da234a299/engine/battle/effects.asm#L194-L212

## **Discord contact info**
graiondilach

### CODE REVIEW

**Comment by GraionDilach** [comment]
> Is the same logic true of sleep and/or poison?

Sleep doesn't have immunity in Gen1. Poison just failed against Poison types (so Tri-Attack in Gen1 couldn't poison Poison types at all). Only these three used the same code path (and all moves capabble of delivering burn/freeze were Ice and Fire moves respectively back then, which wouldn't apply today with cases like Scald).

I can see adding them in, but that wouldn't be Gen1 behaviour.

> Notice also how all those configs have generations as values rather than TRUE/FALSE, so it could be worded in a more neutral way and then the checks need to be against config == GEN_1.

The generational enum in `include/config.h` starts with `GEN_3`. Additionally, `P_SHUCKLE_BERRY_JUICE` in `include/config/pokemon.h` as a Gen2 logic reimplementation is also a binary toggle instead of an enum. As far as I can tell, I'm following existing convention on this front.

**Comment by GraionDilach** [comment]
Turned out Freezing Glare is already in and I just missed it, so this PR is ready for review.

**Comment by Bassoonian** [comment]
Your tests currently fail if the config is set to Gen 1, right? I'd suggest making your tests check for the config and making sure that both settings of the config would pass the test (but not vice versa, if you get what I mean).

**Comment by GraionDilach** [comment]
Updated the tests with Gen1 handling.

**Do not merge this PR as-is until I could rebase out the grabarge commits out from the history.** Ugly commit histories shouldn't be the way to go.

**Comment by GraionDilach** [comment]
Updated for aiming 1.8.0.

File: test/battle/move_effect/burn_hit.c [hunk 1]
```diff
@@ -38,3 +38,32 @@ SINGLE_BATTLE_TEST("Ember cannot burn a Fire-type Pokémon")
         }
     }
 }
+
+#if B_STATUS_TYPE_IMMUNITY > GEN_1
+#define B_STATUS_TYPE_IMMUNITY_TEST "Scald should burn a Water-type Pokémon"
+#else
+#define B_STATUS_TYPE_IMMUNITY_TEST "Scald shouldn't burn a Water-type Pokémon"
+#endif
+SINGLE_BATTLE_TEST(B_STATUS_TYPE_IMMUNITY_TEST)
```
Comment (mrgriffin) [inline test/battle/move_effect/burn_hit.c:47]: I think you may be able to write this a little shorter (and you could drop the `#undef`):
```suggestion
#if B_STATUS_TYPE_IMMUNITY > GEN_1
SINGLE_BATTLE_TEST("Scald should burn a Water-type Pokémon")
#else
SINGLE_BATTLE_TEST("Scald shouldn't burn a Water-type Pokémon")
#endif
```

Comment (GraionDilach) [reply test/battle/move_effect/burn_hit.c]: Thanks, this was reworded.

**Review by Bassoonian — APPROVED** [review]

**Pull request merged**